File: server-python/src/revturbine/core/placements/payload_resolution.py
"""Payload resolution — Python port of
@revt-eng/core/placements/controllers/payload-resolution.ts.

Content lookup + personalization-token application + segment-matched
payload selection. Self-contained: depends only on value types, not on
the engine, providers, or state classes.

The schema content types (``ContentPlacementPayload``, ``MessageBlock``,
``ContentUiPath``, ``ContentPromotion``, ``PersonalizationToken``) are
modeled loosely as ``dict[str, Any]`` here — same approach as the
``PlacementOutput`` alias in helpers.py. The strongly-typed Pydantic
models from ``revturbine_types`` get wired in at TASK-7.

``resolve_payload_for_user_with_provider`` is sync per Q-5 of plan 33;
the TS version is async because provider methods may be remote, but
local-mode providers are CPU-bound. An async ``aresolve_..._with_provider``
variant is a TASK-7 concern.

Source: revturbine-scaffold/src/placements/controllers/payload-resolution.ts
"""

from __future__ import annotations

import re
from typing import Any, Protocol, TypedDict, runtime_checkable

from revturbine.core.helpers import is_record

__all__ = [
    "PayloadResolutionOptions",
    "PersonalizationContext",
    "PlacementContentLookupProvider",
    "ResolvedContent",
    "ResolvedPayload",
    "apply_value_maps",
    "create_static_placement_content_lookup_provider",
    "resolve_content",
    "resolve_payload_for_user",
    "resolve_payload_for_user_with_provider",
    "resolve_tokens",
]


# ── Loose value-type aliases (TASK-7 swaps these for revturbine_types) ───────

PersonalizationContext = dict[str, Any]
ResolvedContent = dict[str, Any]
ContentPlacementPayload = dict[str, Any]
MessageBlock = dict[str, Any]
ContentUiPath = dict[str, Any]
ContentPromotion = dict[str, Any]
PersonalizationToken = dict[str, Any]


class _UserTargetingContext(TypedDict, total=False):
    """The subset of ``UserTargetingContext`` payload resolution reads —
    just the materialized segment IDs.

    Source: core/types.ts ``UserTargetingContext``
    """

    segment_ids: list[str]


class _ResolvedPayloadRequired(TypedDict):
    payload: ContentPlacementPayload
    message_block: MessageBlock
    resolved_content: ResolvedContent


class ResolvedPayload(_ResolvedPayloadRequired, total=False):
    """Source: payload-resolution.ts:18-27"""

    matched_segment_id: str
    ui_path_id: str
    promotion_id: str
    ui_path: ContentUiPath
    promotion: ContentPromotion


class PayloadResolutionOptions(TypedDict, total=False):
    """Source: payload-resolution.ts:128-135"""

    segment_dimensions: dict[str, list[str]]


@runtime_checkable
class PlacementContentLookupProvider(Protocol):
    """Sync lookup provider. The TS protocol allows sync-or-async returns
    (``T | Promise<T>``); the Python port is sync per Q-5.

    Source: payload-resolution.ts:29-35
    """

    def list_payloads(self, surface_template_id: str) -> list[ContentPlacementPayload]: ...
    def get_message_block_by_id(self, block_id: str) -> MessageBlock | None: ...
    def get_ui_path_by_id(self, ui_path_id: str) -> ContentUiPath | None: ...
    def get_promotion_by_id(self, promotion_id: str) -> ContentPromotion | None: ...
    def list_personalization_tokens(self) -> list[PersonalizationToken]: ...


class _StaticLookupProvider:
    """Concrete ``PlacementContentLookupProvider`` built from static lists.

    Source: payload-resolution.ts:37-63
    """

    def __init__(
        self,
        *,
        payloads: list[ContentPlacementPayload] | None = None,
        message_blocks: list[MessageBlock] | None = None,
        ui_paths: list[ContentUiPath] | None = None,
        promotions: list[ContentPromotion] | None = None,
        tokens: list[PersonalizationToken] | None = None,
    ) -> None:
        self._payloads = payloads if isinstance(payloads, list) else []
        self._block_map = {
            b["block_id"]: b
            for b in (message_blocks if isinstance(message_blocks, list) else [])
            if is_record(b) and "block_id" in b
        }
        self._ui_paths = ui_paths if isinstance(ui_paths, list) else []
        self._promotions = promotions if isinstance(promotions, list) else []
        self._tokens = tokens if isinstance(tokens, list) else []

    def list_payloads(self, surface_template_id: str) -> list[ContentPlacementPayload]:
        return [p for p in self._payloads if p.get("surface_template_id") == surface_template_id]

    def get_message_block_by_id(self, block_id: str) -> MessageBlock | None:
        return self._block_map.get(block_id)

    def get_ui_path_by_id(self, ui_path_id: str) -> ContentUiPath | None:
        for item in self._ui_paths:
            if item.get("id") == ui_path_id or item.get("name") == ui_path_id:
                return item
        return None

    def get_promotion_by_id(self, promotion_id: str) -> ContentPromotion | None:
        for item in self._promotions:
            if item.get("id") == promotion_id or item.get("name") == promotion_id:
                return item
        return None

    def list_personalization_tokens(self) -> list[PersonalizationToken]:
        return self._tokens


def create_static_placement_content_lookup_provider(
    *,
    payloads: list[ContentPlacementPayload] | None = None,
    message_blocks: list[MessageBlock] | None = None,
    ui_paths: list[ContentUiPath] | None = None,
    promotions: list[ContentPromotion] | None = None,
    tokens: list[PersonalizationToken] | None = None,
) -> PlacementContentLookupProvider:
    """Build an in-memory ``PlacementContentLookupProvider``.

    Source: payload-resolution.ts:37-63
    """
    return _StaticLookupProvider(
        payloads=payloads,
        message_blocks=message_blocks,
        ui_paths=ui_paths,
        promotions=promotions,
        tokens=tokens,
    )


def _js_string(value: Any) -> str:
    """Match JavaScript's ``String(value)`` coercion for the value kinds
    a personalization context can hold.

    - ``None`` → ``"null"`` (JS ``String(null)``)
    - ``True`` / ``False`` → ``"true"`` / ``"false"``
    - integral ``float`` → no trailing ``.0`` (JS ``String(1.0) === "1"``)
    - everything else → ``str(value)``

    Without this, the parity suite (TASK-8/9/10) would diverge on
    boolean / null / integral-float token substitutions.
    """
    if value is None:
        return "null"
    if value is True:
        return "true"
    if value is False:
        return "false"
    if isinstance(value, float) and value.is_integer():
        return str(int(value))
    return str(value)


def apply_value_maps(
    context: PersonalizationContext,
    tokens: list[PersonalizationToken],
) -> PersonalizationContext:
    """Rewrite token values through each token's ``value_map``.

    A token whose context value (JS-stringified) appears as a key in its
    ``value_map`` is replaced by the mapped value. ``None`` context
    values and tokens without a ``value_map`` are skipped. Mirrors the
    TS ``rawValue == null`` (null/undefined) guard.

    Source: payload-resolution.ts:65-83
    """
    enhanced: PersonalizationContext = dict(context)
    for token_def in tokens:
        token_name = token_def.get("token")
        if token_name is None:
            continue
        raw_value = context.get(token_name)
        value_map = token_def.get("value_map")
        if raw_value is None or not value_map:
            continue
        mapped = value_map.get(_js_string(raw_value))
        if mapped is not None:
            enhanced[token_name] = mapped
    return enhanced


_TOKEN_ALIASES = {
    "current_usage": "usage_current",
    "current_limit": "usage_limit",
    "remaining_usage": "usage_remaining",
}

_TOKEN_RE = re.compile(r"\{\{\s*([a-zA-Z0-9_]+)\s*\}\}")


def resolve_tokens(template: str, context: PersonalizationContext) -> str:
    """Substitute ``{{token_name}}`` placeholders from ``context``.

    Unknown tokens fall back to a small alias table, then are left
    verbatim (the original ``{{...}}`` match) if still unresolved.

    Source: payload-resolution.ts:88-104
    """

    def _replace(match: re.Match[str]) -> str:
        key = match.group(1)
        # TS: `value = context[key]` — absent key is `undefined`. An
        # explicit JS `null` is a present value (String(null) → "null").
        # Python's "absent" is `key not in context`; an explicit None is
        # a present value too. Mirror the TS undefined→alias→leave chain.
        if key in context:
            return _js_string(context[key])
        alias_key = _TOKEN_ALIASES.get(key)
        if alias_key is not None and alias_key in context:
            return _js_string(context[alias_key])
        return match.group(0)  # unresolved → leave the {{token}} verbatim

    return _TOKEN_RE.sub(_replace, template)


def resolve_content(
    content: dict[str, Any],
    context: PersonalizationContext,
) -> ResolvedContent:
    """Resolve token placeholders in every string field of ``content``;
    non-string values pass through untouched.

    Source: payload-resolution.ts:109-119
    """
    resolved: ResolvedContent = {}
    for key, value in content.items():
        resolved[key] = resolve_tokens(value, context) if isinstance(value, str) else value
    return resolved


def _resolve_block_content(
    content: dict[str, Any],
    context: PersonalizationContext,
) -> ResolvedContent:
    """Source: payload-resolution.ts:121-126"""
    return resolve_content(content, context)


def _match_segment_content(
    payload: ContentPlacementPayload,
    user_segment_set: set[str],
    segment_dimensions: dict[str, list[str]] | None,
) -> tuple[str, str | None, str | None, str | None, bool]:
    """Resolve (block_id, segment_id, ui_path_id, promotion_id, skip).

    ``skip`` is True only in the cross-dimension AND path when the user
    fails the AND condition — the caller must skip this payload entirely
    (mirrors the TS ``continue``).

    Source: payload-resolution.ts:156-218
    """
    matched_block_id: str = payload["default_message_block_id"]
    matched_segment_id: str | None = None
    ui_path_id: str | None = payload.get("ui_path_id")
    promotion_id: str | None = payload.get("promotion_id")

    segment_content_map = payload.get("segment_content_map")
    if not segment_content_map:
        return matched_block_id, matched_segment_id, ui_path_id, promotion_id, False

    if segment_dimensions and len(segment_dimensions) > 0:
        # Cross-dimension AND: user must match ≥1 segment per dimension.
        entries_by_dimension: dict[str, list[dict[str, Any]]] = {}
        for entry in segment_content_map:
            dim = entry.get("dimension")
            if not dim:
                continue
            entries_by_dimension.setdefault(dim, []).append(entry)

        if entries_by_dimension:
            all_dimensions_match = True
            first_match_entry: dict[str, Any] | None = None
            for dim_entries in entries_by_dimension.values():
                dim_match = next(
                    (e for e in dim_entries if e["segment_id"] in user_segment_set),
                    None,
                )
                if dim_match is None:
                    all_dimensions_match = False
                    break
                if first_match_entry is None:
                    first_match_entry = dim_match

            if all_dimensions_match and first_match_entry is not None:
                matched_block_id = first_match_entry["message_block_id"]
                matched_segment_id = first_match_entry["segment_id"]
                if first_match_entry.get("ui_path_id"):
                    ui_path_id = first_match_entry["ui_path_id"]
                if first_match_entry.get("promotion_id"):
                    promotion_id = first_match_entry["promotion_id"]
            elif not all_dimensions_match:
                return matched_block_id, matched_segment_id, ui_path_id, promotion_id, True
        else:
            # No dimension metadata — fall back to flat OR.
            for entry in segment_content_map:
                if entry["segment_id"] in user_segment_set:
                    matched_block_id = entry["message_block_id"]
                    matched_segment_id = entry["segment_id"]
                    if entry.get("ui_path_id"):
                        ui_path_id = entry["ui_path_id"]
                    if entry.get("promotion_id"):
                        promotion_id = entry["promotion_id"]
                    break
    else:
        # Flat OR (backward compatible).
        for entry in segment_content_map:
            if entry["segment_id"] in user_segment_set:
                matched_block_id = entry["message_block_id"]
                matched_segment_id = entry["segment_id"]
                if entry.get("ui_path_id"):
                    ui_path_id = entry["ui_path_id"]
                if entry.get("promotion_id"):
                    promotion_id = entry["promotion_id"]
                break

    return matched_block_id, matched_segment_id, ui_path_id, promotion_id, False


def _apply_segment_overrides(
    block: MessageBlock,
    matched_segment_id: str | None,
    user_segment_set: set[str],
) -> tuple[dict[str, Any], str | None]:
    """Merge ``block.default_content`` with a matching segment override and
    return ``(content, matched_segment_id)``.

    When a ``segment_content_map`` entry already selected a segment, honor
    that exact segment. Otherwise (e.g. local mode, where payloads carry a
    single block via ``content_link`` and no ``segment_content_map``) fall
    back to the first block override whose segment matches the user — so
    message-block segment overrides work without a segment→block map in the
    wire format. The matched id is threaded back so the resolved payload
    reflects the override that fired.

    Source: payload-resolution.ts:223-231 (block-override extension, plan 77)
    """
    raw_content: dict[str, Any] = dict(block.get("default_content", {}))
    overrides = block.get("segment_overrides")
    if overrides:
        if matched_segment_id:
            override = next(
                (o for o in overrides if o.get("segment_value_id") == matched_segment_id),
                None,
            )
        else:
            override = next(
                (o for o in overrides if o.get("segment_value_id") in user_segment_set),
                None,
            )
        if override:
            raw_content = {**raw_content, **override.get("content", {})}
            matched_segment_id = override.get("segment_value_id")
    return raw_content, matched_segment_id


def resolve_payload_for_user(
    surface_template_id: str,
    user_context: _UserTargetingContext,
    payloads: list[ContentPlacementPayload],
    blocks: list[MessageBlock],
    tokens: list[PersonalizationToken],
    personalization: PersonalizationContext,
    options: PayloadResolutionOptions | None = None,
) -> ResolvedPayload | None:
    """Resolve the best-matching active payload for a user + surface.

    Source: payload-resolution.ts:137-245
    """
    candidates = [
        p
        for p in payloads
        if p.get("surface_template_id") == surface_template_id and p.get("status") == "active"
    ]
    if not candidates:
        return None

    block_map = {b["block_id"]: b for b in blocks if is_record(b) and "block_id" in b}
    user_segment_set = set(user_context.get("segment_ids") or [])
    enhanced_context = apply_value_maps(personalization, tokens)
    segment_dimensions = options.get("segment_dimensions") if options else None

    for payload in candidates:
        block_id, matched_segment_id, ui_path_id, promotion_id, skip = _match_segment_content(
            payload,
            user_segment_set,
            segment_dimensions,
        )
        if skip:
            continue

        block = block_map.get(block_id)
        if not block or block.get("status") != "active":
            continue

        raw_content, matched_segment_id = _apply_segment_overrides(
            block, matched_segment_id, user_segment_set
        )
        resolved_content = _resolve_block_content(raw_content, enhanced_context)

        result: ResolvedPayload = ResolvedPayload(
            payload=payload,
            message_block=block,
            resolved_content=resolved_content,
        )
        if matched_segment_id is not None:
            result["matched_segment_id"] = matched_segment_id
        if ui_path_id is not None:
            result["ui_path_id"] = ui_path_id
        if promotion_id is not None:
            result["promotion_id"] = promotion_id
        return result

    return None


def resolve_payload_for_user_with_provider(
    surface_template_id: str,
    user_context: _UserTargetingContext,
    provider: PlacementContentLookupProvider,
    personalization: PersonalizationContext,
    explicit_tokens: list[PersonalizationToken] | None = None,
) -> ResolvedPayload | None:
    """Provider-backed variant. Uses flat-OR segment matching only (no
    cross-dimension AND — the TS ``...WithProvider`` path doesn't take
    ``options``). Sync per Q-5; the TS version awaits possibly-async
    provider methods.

    Source: payload-resolution.ts:247-319
    """
    payloads = provider.list_payloads(surface_template_id)
    candidates = [p for p in payloads if p.get("status") == "active"]
    if not candidates:
        return None

    user_segment_set = set(user_context.get("segment_ids") or [])
    provider_tokens = (
        explicit_tokens if explicit_tokens is not None else provider.list_personalization_tokens()
    )
    enhanced_context = apply_value_maps(personalization, provider_tokens)

    for payload in candidates:
        matched_block_id: str = payload["default_message_block_id"]
        matched_segment_id: str | None = None
        ui_path_id: str | None = payload.get("ui_path_id")
        promotion_id: str | None = payload.get("promotion_id")

        segment_content_map = payload.get("segment_content_map")
        if segment_content_map:
            for entry in segment_content_map:
                if entry["segment_id"] in user_segment_set:
                    matched_block_id = entry["message_block_id"]
                    matched_segment_id = entry["segment_id"]
                    if entry.get("ui_path_id"):
                        ui_path_id = entry["ui_path_id"]
                    if entry.get("promotion_id"):
                        promotion_id = entry["promotion_id"]
                    break

        block = provider.get_message_block_by_id(matched_block_id)
        if not block or block.get("status") != "active":
            continue

        raw_content, matched_segment_id = _apply_segment_overrides(
            block, matched_segment_id, user_segment_set
        )
        resolved_content = _resolve_block_content(raw_content, enhanced_context)

        ui_path: ContentUiPath | None = None
        if ui_path_id:
            ui_path = provider.get_ui_path_by_id(ui_path_id)

        promotion: ContentPromotion | None = None
        if promotion_id:
            promotion = provider.get_promotion_by_id(promotion_id)

        result: ResolvedPayload = ResolvedPayload(
            payload=payload,
            message_block=block,
            resolved_content=resolved_content,
        )
        if matched_segment_id is not None:
            result["matched_segment_id"] = matched_segment_id
        if ui_path_id is not None:
            result["ui_path_id"] = ui_path_id
        if promotion_id is not None:
            result["promotion_id"] = promotion_id
        if ui_path is not None:
            result["ui_path"] = ui_path
        if promotion is not None:
            result["promotion"] = promotion
        return result

    return None

File: server-rust/src/placements/payload_resolution.rs
//! Payload resolution: segment matching, block overrides, and token
//! personalization.
//!
//! Given a surface, a user context, and the authored payloads/blocks/tokens,
//! decides **which** content block a user sees and renders its
//! `{{token}}` placeholders.
//!
//! Source: revturbine-scaffold/src/placements/controllers/payload-resolution.ts

use std::collections::{HashMap, HashSet};

use serde_json::{Map, Value};

/// Legacy token spellings, tried when the primary name is absent.
///
/// Source: payload-resolution.ts (TOKEN_ALIASES)
const TOKEN_ALIASES: &[(&str, &str)] = &[
    ("current_usage", "usage_current"),
    ("current_limit", "usage_limit"),
    ("remaining_usage", "usage_remaining"),
];

/// The resolved outcome for one payload.
#[derive(Debug, Clone, PartialEq)]
pub struct ResolvedPayload {
    /// The payload that matched.
    pub payload: Value,
    /// The message block it selected.
    pub message_block: Value,
    /// The block's content with tokens substituted.
    pub resolved_content: Map<String, Value>,
    /// The segment that decided the content, if any.
    pub matched_segment_id: Option<String>,
    /// UI path carried by the payload or its matched segment entry.
    pub ui_path_id: Option<String>,
    /// Promotion carried by the payload or its matched segment entry.
    pub promotion_id: Option<String>,
}

/// Match JavaScript's `String(value)` for the kinds a personalization context
/// can hold.
///
/// Without this the parity suite diverges on boolean, null, and
/// integral-float substitutions: JS renders `null`, `true`, and `1` where a
/// naive Rust `to_string` would render `null`… but `1.0` for the float, and
/// serde's `Value::to_string` would render `"text"` **with quotes** for a
/// string.
///
/// Source: payload-resolution.ts (_js_string)
#[must_use]
pub fn js_string(value: &Value) -> String {
    match value {
        Value::Null => "null".to_string(),
        Value::Bool(true) => "true".to_string(),
        Value::Bool(false) => "false".to_string(),
        // Strings render bare — `Value::to_string` would wrap them in quotes.
        Value::String(s) => s.clone(),
        Value::Number(n) => n.as_f64().map_or_else(
            || n.to_string(),
            |f| {
                if f.fract() == 0.0 && f.is_finite() {
                    // JS `String(1.0) === "1"`.
                    format!("{}", f as i64)
                } else {
                    n.to_string()
                }
            },
        ),
        other => other.to_string(),
    }
}

/// Rewrite token values through each token's `value_map`.
///
/// A token whose context value (JS-stringified) appears as a key in its
/// `value_map` is replaced by the mapped value. Absent/null context values and
/// tokens without a map are skipped.
///
/// Source: payload-resolution.ts:65-83
#[must_use]
pub fn apply_value_maps(context: &Map<String, Value>, tokens: &[Value]) -> Map<String, Value> {
    let mut enhanced = context.clone();
    for token_def in tokens {
        let Some(name) = token_def.get("token").and_then(Value::as_str) else {
            continue;
        };
        let Some(raw) = context.get(name).filter(|v| !v.is_null()) else {
            continue;
        };
        let Some(map) = token_def.get("value_map").and_then(Value::as_object) else {
            continue;
        };
        if let Some(mapped) = map.get(&js_string(raw)) {
            enhanced.insert(name.to_string(), mapped.clone());
        }
    }
    enhanced
}

/// Scan for the next `{{ name }}` starting at `from`.
///
/// Returns `(start, end, name)` with `end` exclusive. Implemented by hand
/// rather than with the `regex` crate: this is one fixed pattern, and a
/// published SDK should not carry `regex-automata` + `aho-corasick` for it.
/// The accepted shape is exactly `\{\{\s*([a-zA-Z0-9_]+)\s*\}\}`.
pub(crate) fn next_token(haystack: &str, from: usize) -> Option<(usize, usize, &str)> {
    let bytes = haystack.as_bytes();
    let mut i = from;

    while i + 1 < bytes.len() {
        if bytes[i] != b'{' || bytes[i + 1] != b'{' {
            i += 1;
            continue;
        }
        let start = i;
        let mut j = i + 2;

        while j < bytes.len() && bytes[j].is_ascii_whitespace() {
            j += 1;
        }
        let name_start = j;
        while j < bytes.len() && (bytes[j].is_ascii_alphanumeric() || bytes[j] == b'_') {
            j += 1;
        }
        let name_end = j;
        while j < bytes.len() && bytes[j].is_ascii_whitespace() {
            j += 1;
        }

        // A name is required, and the braces must close.
        if name_end > name_start && j + 1 < bytes.len() && bytes[j] == b'}' && bytes[j + 1] == b'}'
        {
            return Some((start, j + 2, &haystack[name_start..name_end]));
        }
        // Not a token after all — resume scanning past this `{`.
        i += 1;
    }
    None
}

/// Substitute `{{token}}` placeholders from `context`.
///
/// Unknown names fall back to [`TOKEN_ALIASES`], then are **left verbatim** —
/// an unresolved token renders as its own `{{...}}` rather than as an empty
/// string, so a missing value is visible instead of silently blanking copy.
///
/// Source: payload-resolution.ts:88-104
#[must_use]
pub fn resolve_tokens(template: &str, context: &Map<String, Value>) -> String {
    let mut out = String::with_capacity(template.len());
    let mut cursor = 0usize;

    while let Some((start, end, name)) = next_token(template, cursor) {
        out.push_str(&template[cursor..start]);

        let direct = context.get(name);
        let aliased = || {
            TOKEN_ALIASES
                .iter()
                .find(|(from, _)| *from == name)
                .and_then(|(_, to)| context.get(*to))
        };

        match direct.or_else(aliased) {
            Some(v) => out.push_str(&js_string(v)),
            // Unresolved — keep the literal `{{...}}`.
            None => out.push_str(&template[start..end]),
        }
        cursor = end;
    }

    out.push_str(&template[cursor..]);
    out
}

/// Resolve tokens in every **string** field of `content`; other values pass
/// through untouched.
///
/// Source: payload-resolution.ts:109-119
#[must_use]
pub fn resolve_content(
    content: &Map<String, Value>,
    context: &Map<String, Value>,
) -> Map<String, Value> {
    content
        .iter()
        .map(|(k, v)| {
            let resolved = match v {
                Value::String(s) => Value::String(resolve_tokens(s, context)),
                other => other.clone(),
            };
            (k.clone(), resolved)
        })
        .collect()
}

/// What segment matching decided for one payload.
struct SegmentMatch {
    block_id: String,
    matched_segment_id: Option<String>,
    ui_path_id: Option<String>,
    promotion_id: Option<String>,
    /// True only on the cross-dimension AND path when the user fails the
    /// condition — the caller must skip this payload entirely.
    skip: bool,
}

fn str_of(v: &Value, key: &str) -> Option<String> {
    v.get(key)
        .and_then(Value::as_str)
        .filter(|s| !s.is_empty())
        .map(str::to_string)
}

/// Resolve which block a payload shows for this user.
///
/// Two modes:
///
/// - **Cross-dimension AND** (when `segment_dimensions` is supplied and the
///   entries carry `dimension`): the user must match ≥1 segment in *every*
///   dimension. Failing that skips the payload outright rather than falling
///   back to the default — a payload targeted at "enterprise AND trialing"
///   must not show to someone who is only one of those.
/// - **Flat OR** otherwise: the first matching entry wins.
///
/// Source: payload-resolution.ts:156-218
fn match_segment_content(
    payload: &Value,
    user_segments: &HashSet<String>,
    segment_dimensions: Option<&HashMap<String, Vec<String>>>,
) -> SegmentMatch {
    let mut m = SegmentMatch {
        block_id: payload
            .get("default_message_block_id")
            .and_then(Value::as_str)
            .unwrap_or_default()
            .to_string(),
        matched_segment_id: None,
        ui_path_id: str_of(payload, "ui_path_id"),
        promotion_id: str_of(payload, "promotion_id"),
        skip: false,
    };

    let Some(entries) = payload
        .get("segment_content_map")
        .and_then(Value::as_array)
        .filter(|a| !a.is_empty())
    else {
        return m;
    };

    let mut take = |entry: &Value| {
        if let Some(id) = entry.get("message_block_id").and_then(Value::as_str) {
            m.block_id = id.to_string();
        }
        m.matched_segment_id = str_of(entry, "segment_id");
        // Entry-level path/promotion override the payload's only when present.
        if let Some(p) = str_of(entry, "ui_path_id") {
            m.ui_path_id = Some(p);
        }
        if let Some(p) = str_of(entry, "promotion_id") {
            m.promotion_id = Some(p);
        }
    };

    let dimensioned = segment_dimensions.is_some_and(|d| !d.is_empty());
    if dimensioned {
        let mut by_dimension: HashMap<&str, Vec<&Value>> = HashMap::new();
        for e in entries {
            if let Some(dim) = e
                .get("dimension")
                .and_then(Value::as_str)
                .filter(|d| !d.is_empty())
            {
                by_dimension.entry(dim).or_default().push(e);
            }
        }

        if !by_dimension.is_empty() {
            let mut all_match = true;
            let mut first: Option<&Value> = None;
            for dim_entries in by_dimension.values() {
                let hit = dim_entries.iter().find(|e| {
                    e.get("segment_id")
                        .and_then(Value::as_str)
                        .is_some_and(|s| user_segments.contains(s))
                });
                match hit {
                    Some(e) => {
                        if first.is_none() {
                            first = Some(e);
                        }
                    }
                    None => {
                        all_match = false;
                        break;
                    }
                }
            }

            if all_match {
                if let Some(e) = first {
                    take(e);
                }
            } else {
                m.skip = true;
            }
            return m;
        }
        // No dimension metadata on the entries — fall through to flat OR.
    }

    for e in entries {
        if e.get("segment_id")
            .and_then(Value::as_str)
            .is_some_and(|s| user_segments.contains(s))
        {
            take(e);
            break;
        }
    }
    m
}

/// Merge a block's `default_content` with a matching segment override.
///
/// When segment matching already chose a segment, that exact override is
/// honoured. Otherwise the first override whose segment the user holds wins —
/// which is what makes block-level overrides work in local mode, where
/// payloads carry a single block and no `segment_content_map`.
///
/// Source: payload-resolution.ts:223-231
fn apply_segment_overrides(
    block: &Value,
    matched_segment_id: Option<String>,
    user_segments: &HashSet<String>,
) -> (Map<String, Value>, Option<String>) {
    let mut content = block
        .get("default_content")
        .and_then(Value::as_object)
        .cloned()
        .unwrap_or_default();

    let Some(overrides) = block.get("segment_overrides").and_then(Value::as_array) else {
        return (content, matched_segment_id);
    };

    let found = match matched_segment_id.as_deref() {
        Some(id) => overrides
            .iter()
            .find(|o| o.get("segment_value_id").and_then(Value::as_str) == Some(id)),
        None => overrides.iter().find(|o| {
            o.get("segment_value_id")
                .and_then(Value::as_str)
                .is_some_and(|s| user_segments.contains(s))
        }),
    };

    let Some(over) = found else {
        return (content, matched_segment_id);
    };

    if let Some(patch) = over.get("content").and_then(Value::as_object) {
        for (k, v) in patch {
            content.insert(k.clone(), v.clone());
        }
    }
    let id = str_of(over, "segment_value_id").or(matched_segment_id);
    (content, id)
}

/// Resolve the best-matching **active** payload for a user and surface.
///
/// Candidates are filtered to the surface and to `status == "active"`, then
/// tried in authored order; the first that yields an active block wins. A
/// payload whose block is missing or inactive is skipped rather than
/// rendered empty.
///
/// Source: payload-resolution.ts:137-245
#[must_use]
pub fn resolve_payload_for_user(
    surface_template_id: &str,
    user_segment_ids: &[String],
    payloads: &[Value],
    blocks: &[Value],
    tokens: &[Value],
    personalization: &Map<String, Value>,
    segment_dimensions: Option<&HashMap<String, Vec<String>>>,
) -> Option<ResolvedPayload> {
    let candidates: Vec<&Value> = payloads
        .iter()
        .filter(|p| {
            p.get("surface_template_id").and_then(Value::as_str) == Some(surface_template_id)
                && p.get("status").and_then(Value::as_str) == Some("active")
        })
        .collect();
    if candidates.is_empty() {
        return None;
    }

    let block_map: HashMap<&str, &Value> = blocks
        .iter()
        .filter(|b| b.is_object())
        .filter_map(|b| b.get("block_id").and_then(Value::as_str).map(|id| (id, b)))
        .collect();

    let user_segments: HashSet<String> = user_segment_ids.iter().cloned().collect();
    let enhanced = apply_value_maps(personalization, tokens);

    for payload in candidates {
        let m = match_segment_content(payload, &user_segments, segment_dimensions);
        if m.skip {
            continue;
        }

        let Some(block) = block_map.get(m.block_id.as_str()) else {
            continue;
        };
        if block.get("status").and_then(Value::as_str) != Some("active") {
            continue;
        }

        let (raw_content, matched_segment_id) =
            apply_segment_overrides(block, m.matched_segment_id, &user_segments);

        return Some(ResolvedPayload {
            payload: (*payload).clone(),
            message_block: (*block).clone(),
            resolved_content: resolve_content(&raw_content, &enhanced),
            matched_segment_id,
            ui_path_id: m.ui_path_id,
            promotion_id: m.promotion_id,
        });
    }

    None
}
